[Profiling] Allow to upgrade managed ILM policy

danielmitterdorfer opened the pull request:
With this commit we amend the default lifecycle policy for Universal Profiling with a version field in the metadata section and add infrastructure to determine whether a lifecycle policy should be upgraded to a newer version.

danielmitterdorfer committed:
[Profiling] Allow to upgrade managed ILM policy

With this commit we amend the default lifecycle policy for Universal
Profiling with a version field in the metadata section and add
infrastructure to determine whether a lifecycle policy should be
upgraded to a newer version.

elasticsearchmachine commented:
Pinging @elastic/profiling (Team:Universal Profiling)

elasticsearchmachine commented:
Hi @danielmitterdorfer, I've created a changelog YAML for you.

danielmitterdorfer committed:
Update docs/changelog/96550.yaml

danielmitterdorfer commented:
@gmarouli I've assigned you as reviewer as this PR also touches `IndexTemplateRegistry`.

danielmitterdorfer committed:
Remove unused import

gmarouli reviewed (commented):
Looks very good, almost done, I added some small comments. Let me know what you think :).

danielmitterdorfer committed:
Address review comments

danielmitterdorfer commented on x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexTemplateRegistry.java:
```diff
@@ -214,6 +219,22 @@ protected Map<String, ComposableIndexTemplate> getComposableTemplateConfigs() {
+    protected boolean isUpgradeRequired(LifecyclePolicy currentPolicy, LifecyclePolicy newPolicy) {
+        try {
+            return getVersion(currentPolicy) < getVersion(newPolicy);
+        } catch (NumberFormatException ex) {
```
I've restructured the code now in https://github.com/elastic/elasticsearch/pull/96550/commits/a90bf4314558810caeb4f8dd30f2133f5280cc56. To avoid complicating the control flow I kept the `catch` block here but raise a specific exception now on parse errors in `#getVersion()`.

danielmitterdorfer commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java:
```diff
@@ -488,19 +488,20 @@ public void onFailure(Exception e) {
+                if (Objects.isNull(currentPolicy)) {
                     logger.debug("adding lifecycle policy [{}] for [{}], because it doesn't exist", policy.getName(), getOrigin());
                     putPolicy(policy, creationCheck);
+                } else if (isUpgradeRequired(currentPolicy, policy)) {
```
I've amended test cases for the base class in https://github.com/elastic/elasticsearch/pull/96550/commits/a90bf4314558810caeb4f8dd30f2133f5280cc56.

gmarouli replied:
> Thanks for the review @gmarouli! I've addressed your feedback in [a90bf43](https://github.com/elastic/elasticsearch/commit/a90bf4314558810caeb4f8dd30f2133f5280cc56). Could you please take another look?

Yes, on it.

danielmitterdorfer commented:
Thanks for the review @gmarouli! I've addressed your feedback in https://github.com/elastic/elasticsearch/pull/96550/commits/a90bf4314558810caeb4f8dd30f2133f5280cc56. Could you please take another look?

danielmitterdorfer commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'main' into profiling-updateable-policy

gmarouli reviewed (approved):
LGTM, thank you for addressing the comments! 🚀
